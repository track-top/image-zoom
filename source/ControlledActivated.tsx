import React, {
  FC,
  ReactNode,
  RefObject,
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'
import { createPortal } from 'react-dom'
import { useEvent, usePrevious, useWindowSize } from 'react-use'
import {
  getModalContentStyle,
  getModalOverlayStyle,
  pseudoParentEl,
} from './helpers'

const rmizPortalEl = document.createElement(`div`)
document.body.appendChild(rmizPortalEl)

export interface ControlledActivatedProps {
  children: ReactNode
  closeText?: string
  isActive: boolean
  onLoad: () => void
  onUnload: () => void
  onZoomChange?: (value: boolean) => void
  overlayBgColorEnd?: string
  overlayBgColorStart?: string
  parentRef: RefObject<HTMLElement>
  portalEl?: HTMLElement
  scrollableEl?: HTMLElement | Window
  transitionDuration?: number
  zoomMargin?: number
  zoomZindex?: number
}

const ControlledActivated: FC<ControlledActivatedProps> = ({
  children,
  closeText = `Unzoom Image`,
  isActive: isActiveFromParent,
  onLoad,
  onUnload,
  onZoomChange,
  overlayBgColorEnd = `rgba(255, 255, 255, 0.95)`,
  overlayBgColorStart = `rgba(255, 255, 255, 0)`,
  parentRef,
  portalEl = rmizPortalEl,
  scrollableEl = window,
  transitionDuration = 300,
  zoomMargin = 0,
  zoomZindex = 2147483647,
}: ControlledActivatedProps) => {
  const btnRef = useRef<HTMLButtonElement>(null)
  const [, forceUpdate] = useState<number>(0)
  const [isActive, setIsActive] = useState<boolean>(isActiveFromParent)
  const [isLoaded, setIsLoaded] = useState<boolean>(false)
  const [isUnloading, setIsUnloading] = useState<boolean>(false)
  const prevIsActive = usePrevious(isActive)
  const prevIsActiveFromParent = usePrevious(isActiveFromParent)
  const prevIsLoaded = usePrevious(isLoaded)
  const { width: innerWidth, height: innerHeight } = useWindowSize()

  // on click, tell caller it should zoom
  const handleClick = useCallback(
    e => {
      e.preventDefault()
      if (onZoomChange) {
        onZoomChange(false)
      }
    },
    [onZoomChange]
  )

  // on escape, tell caller it should unzoom
  const handleKeyDown = useCallback(
    e => {
      if (isActive && (e.key === `Escape` || e.keyCode === 27)) {
        e.stopPropagation()
        if (onZoomChange) {
          onZoomChange(false)
        }
      }
    },
    [isActive, onZoomChange]
  )

  const handleScroll = useCallback(() => {
    forceUpdate(n => n + 1)

    if (!isUnloading && onZoomChange) {
      onZoomChange(false)
    }
  }, [isUnloading, onZoomChange])

  // listen for keydown on the document
  useEvent(`keydown`, handleKeyDown, document)

  // listen for scroll and close
  useEvent(`scroll`, handleScroll, scrollableEl)

  // set loaded on mount and focus
  useEffect(() => {
    if (!prevIsActive && isActive) {
      setIsLoaded(true)

      if (btnRef.current) {
        btnRef.current.focus({ preventScroll: true })
      }
    }
  }, [isActive, prevIsActive])

  useEffect(() => {
    // when parent says to deactivate, begin unloading process
    if (prevIsActiveFromParent && !isActiveFromParent) {
      setIsUnloading(true)
    }

    // when parent says to activate, begin active process
    if (!prevIsActiveFromParent && isActiveFromParent) {
      setIsActive(true)
    }
  }, [isActiveFromParent, prevIsActiveFromParent])

  // if unloading, tell parent that we're all done here after Nms
  useEffect(() => {
    let unloadTimeout: NodeJS.Timeout

    if (isUnloading) {
      unloadTimeout = setTimeout(() => {
        setIsLoaded(false)
        setIsActive(false)
        setIsUnloading(false)
      }, transitionDuration)
    }

    return (): void => {
      clearTimeout(unloadTimeout)
    }
  }, [isUnloading, transitionDuration])

  // let parent know of changes to load status
  useEffect(() => {
    if (!prevIsLoaded && isLoaded) {
      onLoad()
    }
    if (prevIsLoaded && !isLoaded) {
      onUnload()
    }
  }, [isLoaded, onLoad, onUnload, prevIsLoaded])

  // use parent element or fake one if it's not yet loaded
  const parentEl = parentRef.current || pseudoParentEl

  // get parent element's dimensions
  const { height, left, top, width } = parentEl.getBoundingClientRect()

  const overlayStyle = getModalOverlayStyle({
    isLoaded,
    isUnloading,
    overlayBgColorEnd,
    overlayBgColorStart,
    transitionDuration,
    zoomZindex,
  })

  const contentStyle = getModalContentStyle({
    height,
    isLoaded,
    innerHeight,
    innerWidth,
    isUnloading,
    left,
    originalTransform: parentEl.style.transform,
    top,
    transitionDuration,
    width,
    zoomMargin,
  })

  return isActive
    ? createPortal(
      <div aria-modal data-rmiz-overlay role="dialog" style={overlayStyle}>
        <div data-rmiz-modal-content style={contentStyle}>
          {children}
        </div>
        <button
          aria-label={closeText}
          data-rmiz-btn-close
          onClick={handleClick}
          ref={btnRef}
          type="button"
        />
      </div>,
      portalEl
    )
    : null
}

export default memo(ControlledActivated)
