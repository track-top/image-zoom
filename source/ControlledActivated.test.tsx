import { render } from '@testing-library/react'
import React, { createRef } from 'react'
import ControlledActivated from './ControlledActivated'

test(`uses pseudo-parent when no parentRef.current`, async () => {
  render(
    <ControlledActivated
      closeText="Unzoom image"
      isActive
      onLoad={jest.fn()}
      onUnload={jest.fn()}
      onZoomChange={jest.fn()}
      overlayBgColorEnd="rgba(255, 255, 255, 0.95)"
      overlayBgColorStart="rgba(255, 255, 255, 0)"
      parentRef={createRef()}
      portalEl={document.body}
      scrollableEl={window}
      transitionDuration={300}
      zoomMargin={0}
      zoomZindex={2147483647}
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledActivated>
  )

  expect(document.body).toMatchSnapshot()
})

test(`render with default props`, async () => {
  render(
    <ControlledActivated
      isActive
      onLoad={jest.fn()}
      onUnload={jest.fn()}
      parentRef={createRef()}
      portalEl={document.body}
      scrollableEl={window}
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </ControlledActivated>
  )

  expect(document.body).toMatchSnapshot()
})
