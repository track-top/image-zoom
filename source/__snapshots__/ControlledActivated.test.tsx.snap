// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with default props 1`] = `
<body>
  <div />
  <div />
  <div
    aria-modal="true"
    data-rmiz-overlay="true"
    role="dialog"
    style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
  >
    <div
      data-rmiz-modal-content="true"
      style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
    </div>
    <button
      aria-label="Unzoom Image"
      data-rmiz-btn-close="true"
      type="button"
    />
  </div>
</body>
`;

exports[`uses pseudo-parent when no parentRef.current 1`] = `
<body>
  <div />
  <div />
  <div
    aria-modal="true"
    data-rmiz-overlay="true"
    role="dialog"
    style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
  >
    <div
      data-rmiz-modal-content="true"
      style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
    </div>
    <button
      aria-label="Unzoom image"
      data-rmiz-btn-close="true"
      type="button"
    />
  </div>
</body>
`;
